client/client.go

```diff
@@ -1,71 +1,378 @@
+	}
 
-	// if no meta, get root.json
-	// if meta, get timestamp.json
-	// if new, get snapshot.json
-	// if new root.json, restart
-	// if new targets.json update
-	// fully check all signatures
+	if rootJSON, ok := meta["root.json"]; ok {
+		// unmarshal root.json without verifying as we need the root
+		// keys first
+		s := &data.Signed{}
+		if err := json.Unmarshal(rootJSON, s); err != nil {
+			return err
+		}
+		root := &data.Root{}
+		if err := json.Unmarshal(s.Signed, root); err != nil {
+			return err
+		}
+		db := keys.NewDB()
+		for id, k := range root.Keys {
+			if err := db.AddKey(id, k); err != nil {
+				return err
+			}
+		}
+		for name, role := range root.Roles {
+			if err := db.AddRole(name, role); err != nil {
+				return err
+			}
+		}
+		if err := signed.Verify(s, "root", 0, db); err != nil {
+			return err
+		}
+		c.root = root
+		c.db = db
+	} else {
+		return ErrNoRootKeys
+	}
 
-	/*
-		If at any point in the following process there is a problem (e.g., only expired
-		metadata can be retrieved), the Root file is downloaded and the process starts
-		over. Optionally, the software update system using the framework can decide how
-		to proceed rather than automatically downloading a new Root file.
+	if snapshotJSON, ok := meta["snapshot.json"]; ok {
+		snapshot := &data.Snapshot{}
+		if err := signed.Unmarshal(snapshotJSON, snapshot, "snapshot", 0, c.db); err != nil {
+			return err
+		}
+		c.snapshot = snapshot
+	}
 
-			TUF downloads and verifies timestamp.json.
+	if targetsJSON, ok := meta["targets.json"]; ok {
+		targets := &data.Targets{}
+		if err := signed.Unmarshal(targetsJSON, targets, "targets", 0, c.db); err != nil {
+			return err
+		}
+		c.targets = targets
+	}
 
-			If timestamp.json indicates that snapshot.json has changed, TUF downloads
-			and verifies snapshot.json.
+	if timestampJSON, ok := meta["timestamp.json"]; ok {
+		timestamp := &data.Timestamp{}
+		if err := signed.Unmarshal(timestampJSON, timestamp, "timestamp", 0, c.db); err != nil {
+			return err
+		}
+		c.timestamp = timestamp
+	}
 
-			TUF determines which metadata files listed in snapshot.json differ from
-			those described in the last snapshot.json that TUF has seen. If root.json
-			has changed, the update process starts over using the new root.json.
+	c.localMeta = meta
+	return nil
+}
 
-			TUF provides the software update system with a list of available files
-			according to targets.json.
-	*/
+// downloadMeta downloads top-level metadata from remote storage and verifies
+// it using the given file metadata.
+func (c *Client) downloadMeta(name string, m *data.FileMeta) ([]byte, error) {
+	r, err := c.remote.Get(name)
+	if err != nil {
+		if err == ErrNotFound {
+			return nil, ErrMissingRemoteMetadata{name}
+		}
+		return nil, err
+	}
+	defer r.Close()
+
+	// if m is nil (e.g. when downloading timestamp.json, which has unknown
+	// size), just read the entire stream and return it
+	if m == nil {
+		b, err := ioutil.ReadAll(r)
```

Comment:
Is there a stdlib method to read at most N bytes? Just checking before I implement it.